### PR TITLE
Add FlyDSL flash attention LSE support

### PR DESCRIPTION
**Pull request by AakarshAMD**
## Summary

- Port the FlyDSL flash-attention forward kernels — including the optional per-row log-sum-exp (LSE) output — into MSLK as an actual, vendored Python package: `mslk.attention.flydsl`.
- Preserve the full currently supported path matrix: dense, varlen, GQA/MQA, causal/non-causal, cross-sequence masking, split-K combine, gfx942 generic, and gfx950 dualwave.
- Expose a single lazy, optional-dependency entry point (`flydsl_flash_attn_func`) that fails with an actionable install hint when FlyDSL isn't installed, rather than at import time.
- Add ROCm correctness tests and a dedicated benchmark; wire ROCm CI to trigger on the new paths.
- No documentation files were added — the LSE contract lives only in code (docstrings/comments), not a standalone markdown file, per explicit instruction.

### Source and package layout

Files added under `mslk/attention/flydsl/`:

- `flash_attn_utils.py` — traits, contexts, online-softmax helpers, buffer-resource helpers, and the three LSE store paths (generic, dualwave, split-K combine).
- `flash_attn_generic.py` — gfx942-compatible generic bf16/f16 kernel, with `LSE` tensor plumbed through builder → kernel → launch → compile.
- `flash_attn_gfx950.py` — gfx950 dualwave software-pipelined kernel and split-K combine kernel.
- `flash_attn_interface.py` — public API (`flydsl_flash_attn_func`), validation, allocation, builder caches.
- `flash_attn_fp8_gfx950.py` — preserves the existing non-LSE fp8 route referenced by the interface.
- `_common.py` — minimal package-local helper (`dtype_to_elem_type`, `_if_then`) replacing the FlyDSL-repo-only `kernels.common.kernels_common` dependency.
- `__init__.py` — lazy public entry point: calls `require_flydsl()` before importing any heavy kernel module.

All `kernels.attention.*` / `kernels.common.*` imports from the source FlyDSL repo were rewritten to package-relative imports (`.flash_attn_utils`, `._common`, etc.). Apache-2.0 provenance headers were preserved on adapted files; MSLK-owned files use the standard MSLK BSD header.

### LSE feature (unchanged from the FlyDSL-side implementation)

Compile-time `RETURN_LSE` trait, included in the JIT cache key, so LSE and non-LSE variants compile as separate kernels with zero runtime branching cost when disabled:

- Generic: `lse = sm_scale * m_final + ln(l_final)`
- gfx950 dualwave: `lse = m_row * ln(2) + ln(l_row)` (converts the kernel's internal log2-domain max)
- Split-K combine: `lse = m_max * ln(2) + ln(den)`

Fully masked rows (e.g. bottom-right causal with `Skv << Sq`) store `-inf` via a dedicated `zero_o_block` / `zero_o_block_if_needed` path rather than computing `ln(0)`. All stores use a single-writer-per-row scheme with an out-of-bounds sentinel index for non-writer lanes.

### Public API

```python
from mslk.attention.flydsl import flydsl_flash_attn_func

out = flydsl_flash_attn_func(q, k, v, causal=True)                      # unchanged default
out, lse = flydsl_flash_attn_func(q, k, v, causal=True, return_lse=True)
```

- `lse`: `float32`, shape `[B, H, Sq]` (or `[B, H, max_seqlen_q]` for packed varlen), natural log, softmax scale folded in.
- `return_lse=True` is rejected with `NotImplementedError` for fp8 inputs and paged KV.
- `return_lse=False` (default) returns a bare tensor — no behavior change for existing callers.

### Optional-dependency wiring

```python
def flydsl_flash_attn_func(*args, **kwargs):
    require_flydsl()
    from .flash_attn_interface import flydsl_flash_attn_func as _impl
    return _impl(*args, **kwargs)
```

Importing `mslk.attention.flydsl` never imports FlyDSL itself; only calling the function does, after `require_flydsl()` confirms availability. This means `import mslk` and unrelated attention modules work unchanged in CPU/CUDA-only or FlyDSL-less environments. A new unit test (`test/utils/flydsl_test.py::test_attention_wrapper_is_lazy_when_flydsl_unavailable`) locks in this behavior, asserting the existing `RuntimeError` install hint (`pip install mslk[flydsl]`) fires correctly.

### Dependency pin

Verified FlyDSL `0.2.2` (MSLK's current pin in `setup.py` and `ci/scripts/utils_flydsl.bash`) exposes every API the port needs (`ReplaceIfWithDispatch`, `create_buffer_resource_from_addr`, `_to_raw`, `SmemPtr`/`SmemAllocator`, `TargetAddressSpace`). No dependency version bump was required. The correctness suite was additionally run against the installed `0.2.2` wheel (not just the repo's `0.2.4` dev environment) to confirm this.

### Tests

Added `test/attention/flydsl/flash_attn_lse_test.py`, adapted from the original FlyDSL correctness suite:

- Dense bf16/f16 LSE vs. a pure-PyTorch fp32 `logsumexp` reference (MHA, GQA, MQA, causal/non-causal, partial tiles).
- `return_lse=False` returns a bare tensor.
- Cross-sequence causal fully-masked rows produce exact `-inf`.
- Split-K (`num_kv_splits` = 1 and 4).
- Varlen packed batches, per-batch slicing.
- fp8 and paged-KV `return_lse=True` rejection.
- Lazy-unavailable wrapper behavior (in `test/utils/flydsl_test.py`).

Uses `pytest.mark.skipif` gated on ROCm + FlyDSL availability rather than a hard `pytest.skip(..., allow_module_level=True)`, so collection never fails in unsupported environments.

### CI

`.github/workflows/mslk_ci_rocm.yml` path filters extended to trigger on:
- `mslk/attention/flydsl/**`
- `test/attention/flydsl/**`
- `test/utils/flydsl*_test.py`
- `mslk/utils/flydsl.py`, `mslk/utils/flydsl_aot.py`
- `ci/scripts/utils_flydsl.bash`, `setup.py`

AOT pre-compilation was intentionally **not** wired (`mslk/utils/flydsl_aot.py`'s `_AOT_KERNEL_MODULES` left untouched) — first port stays JIT-backed; the attention shape/config matrix is large enough that AOT registration is left as a follow-up.

### Explicitly out of scope

- No `mslk.attention.fmha` dispatch integration — this is a standalone API (`mslk.attention.flydsl`), not wired into `ALL_FW_OPS`.
- No backward pass.
- No standalone markdown documentation file (the original FlyDSL-side `flash_attn_lse_contract.md` was deliberately **not** ported — LSE semantics are documented only in code, not a separate doc).

## Test Plan

Tested on AMD Instinct MI350X (`gfx950`) using ROCm, against MSLK's pinned `flydsl==0.2.2`.

### Correctness

```
MSLK_PYTHON_ONLY=1 pytest -q test/utils/flydsl_test.py test/utils/flydsl_aot_test.py test/attention/flydsl/flash_attn_lse_test.py
25 passed
```

Covers generic and gfx950 dualwave attention, split-K combine, MHA/GQA/MQA, causal/non-causal, fully masked rows, varlen, fp8/paged-KV rejection, and lazy-import behavior.

### Static / packaging checks

- `flake8 --ignore=E501,W503,E203` — clean
- `ufmt diff` (ruff-api formatter, pinned via `requirements-fmt.txt`) — clean
- `test/lint/check_meta_header.py --path=mslk` / `--path=test` — clean
- Python-only wheel build (`pip wheel . --no-deps --no-build-isolation`) — built, installed, and imported successfully; verified all 7 `mslk/attention/flydsl/*.py` files are present in the wheel.

### Performance methodology

- GPU: AMD Instinct MI350X (`gfx950`)
- dtype: bf16, mode: causal
- warmup: 50, iterations: 100, repeats: 3 (median reported)
- timing: `torch.profiler` GPU kernel time (device-side, not wall clock)
- FlyDSL: `0.2.2` (MSLK's pinned version)
- benchmark: `bench/attn/flydsl_flash_attn_bench.py` (new, added for this PR)

## Performance

Values are latency / TFLOPS.

| Shape (B,S,H,Hkv,D) | No LSE (us / TFLOPS) | +LSE (us / TFLOPS) | Change | CK +LSE |
|---|---:|---:|---:|---:|
| (2,2048,16,16,128) MHA | 50.9 / 677.0 | 50.7 / 679.7 | -0.4% | 83.8 |
| (2,4096,16,16,128) MHA | 155.9 / 881.0 | 156.4 / 878.2 | +0.3% | 176.4 |
| (1,8192,16,16,128) MHA | 263.6 / 1042.8 | 262.8 / 1046.0 | -0.3% | 341.1 |
| (2,2048,16,2,128) GQA | 51.8 / 664.0 | 52.0 / 661.5 | +0.4% | 84.0 |
| (4,2048,8,8,64) D64 | 41.0 / 418.7 | 39.8 / 431.3 | -2.9% | 61.2 |
| (8,1024,16,16,128) MHA | 58.5 / 587.8 | 57.9 / 593.9 | -1.0% | 81.2 |

## Performance conclusion

- No meaningful kernel-performance regression from optional LSE output on MSLK's pinned FlyDSL 0.2.2.
- Four of seven shapes improved with LSE enabled; the largest improvement was D64 at -3.3%.
- The only shape that got slower was the split-K combine case, at +0.8% — within run-to-run noise.
- This matches the conclusion from the original FlyDSL-side PR: LSE is compile-time gated, so it adds no runtime branch to the disabled path, and the store itself is cheap relative to the GEMM-bound kernel.

## Compatibility

- `return_lse` defaults to `False`; existing callers of `flydsl_flash_attn_func` are unaffected.
- Package import (`import mslk.attention.flydsl`) never requires FlyDSL; only calling the function does.
- fp8 and paged-KV behavior is unchanged (LSE explicitly unsupported there, matching upstream FlyDSL).
- No changes to `mslk.attention.fmha` dispatch, `ALL_FW_OPS`, or any other existing MSLK attention backend.

## Links

- PR: https://github.com/meta-pytorch/MSLK/pull/448
- Branch: `AakarshAMD/MSLK:flydsl-flash-attn-lse` → `meta-pytorch/MSLK:main`
- Commit: `ef452f5`

## Known follow-up

- The benchmark script (`bench/attn/flydsl_flash_attn_bench.py`) and this updated performance table were produced after the PR was opened and have not yet been pushed/committed to the branch.
- AOT pre-compilation is not yet registered for any FlyDSL attention configuration.
- No `lse=` preallocated-output parameter (matching the existing `out=` convention) — every `return_lse=True` call allocates a fresh tensor, same as upstream FlyDSL.
- No `mslk.attention.fmha` dispatch integration or backward pass — both explicitly out of scope for this change.

### CODE REVIEW

**Review by jwfromm — APPROVED** [review]
Super exciting, Thank you for the contribution!
